### PR TITLE
AP_Compass: reset compass ids not present after compass cal

### PR DESCRIPTION
**Pull request by bugobliterator**

### CODE REVIEW

**Comment by rmackay9** [comment]
@bugobliterator, thanks for this.  Can you tell us what testing you've done?  Recently we're asking submitters of PRs to explain what testing they've done.

**Comment by rmackay9** [comment]
Hmm.. perhaps not much!  Looks like it's failing travis with this error, "_priority_did_stored_list[i].set_and_save(0);".

**Comment by khancyr** [comment]
and you could totally add a check for this on the test suuite for compass !

**Comment by bugobliterator** [comment]
> @bugobliterator, thanks for this. Can you tell us what testing you've done? Recently we're asking submitters of PRs to explain what testing they've done.

Hi @rmackay9 , Sorry forgot to mention. Still getting tested through Mission Planner. Will let you know once its done, and post results here, when ready for review.

**Comment by meee1** [comment]
image
new prompt when detected a missing ID when it has a prio slot
image
and the "remove missing" button as well as the missing column to show whats not detected.

i verified sids change by adding a uavcan compass, rebooting with it removed, MP told me it was missing, i ignored it and did a new onboard compass cal, and once finished only the compass's left where detect/calibrated while removing the uavcan compass i had removed

**Comment by tridge** [comment]
I think doing it after the cal doesn't fix the issue fully. Imagine the user had 3 compasses, then swaps out the first one. If we do the reset after the cal then the new compass won't be calibrated, as it will be past slot 3
I think we need to do this before the cal.

**Comment by bugobliterator** [comment]
> I think doing it after the cal doesn't fix the issue fully. Imagine the user had 3 compasses, then swaps out the first one. If we do the reset after the cal then the new compass won't be calibrated, as it will be past slot 3
> I think we need to do this before the cal.

Yeah, agreed. I will make the change and push again.

**Comment by rmackay9** [comment]
.. and it still makes sense to do it first if someone cancels the compass calibration before it finishes?

**Comment by bugobliterator** [comment]
> .. and it still makes sense to do it first if someone cancels the compass calibration before it finishes? I'm just thinking that users may think that cancelling a compass calibration will mean that nothing will have changed.. but something will have..

The user had a chance to not proceed when the dialog from MP is shown, so I believe it should be ok. Maybe a bit more info saying "Note: Clicking OK will reconfigure your Compass settings." or something like that might help. I will try this out and work with Michael.

**Comment by rmackay9** [comment]
@bugobliterator, ok great.  If we can't think of a reason why the user might care about the extra compasses being removed

**Comment by rmackay9** [comment]
Oops, sorry, I tried to cancel a comment and accidentally closed (but then re-opened) the PR.  anyway, sounds good.

**Comment by bugobliterator** [comment]
@rmackay9 @tridge 
I have made the changes as requested. Now, after calibration request we remove undetected compasses and run compass init() again, so as to allocate replaced compass. Various combinations of 2 internal one external I2C and one external CAN Compasses were tested. Calibrating the compass after removal went smoothly and user can continue operation, as normal. If there was a replacement compass connected before cal, it will replace the old undetected one and calibration may proceed.

Tests with Mission Planner will be done by Michael and then you can give it a try as well.

**Comment by khancyr** [comment]
could you add this to the test suite ?

**Comment by bugobliterator** [comment]
> could you add this to the test suite ?

I will, once it is approved.

File: libraries/AP_Compass/AP_Compass_Calibration.cpp [hunk 1]
```diff
@@ -370,7 +370,9 @@ MAV_RESULT Compass::handle_mag_cal_command(const mavlink_command_long_t &packet)
         bool autosave = !is_zero(packet.param3);
         float delay = packet.param4;
         bool autoreboot = !is_zero(packet.param5);
-
+        // Before we begin lets refresh our existing list of compasses
+        _reset_compass_id();
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass_Calibration.cpp:374]: this is also needed in Compass::mag_cal_fixed_yaw()

File: libraries/AP_Compass/AP_Compass.cpp [hunk 2]
```diff
@@ -1620,6 +1623,30 @@ bool Compass::configured(char *failure_msg, uint8_t failure_msg_len)
...
+            _priority_did_stored_list[i].set_and_save(0);
+        }
+    }
+    
+    // Check if any of the registered devs are not registered
+    for (StateIndex i(0); i<COMPASS_MAX_INSTANCES; i++) {
+        if (_state[i].dev_id == 0) {
+            // also hard reset dev_ids that are not detected
+            _state[i].dev_id.save();
+            gcs().send_text(MAV_SEVERITY_ALERT, "Mag: Compass #%d with DEVID %lu deregistered!", uint8_t(i), (unsigned long)_state[i].expected_dev_id);
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass.cpp:1641]: should not do this if _state[i].expected_dev_id is zero. I see:
APM: Mag: Compass #2 with DEVID 0 deregistered!

File: libraries/AP_Compass/AP_Compass.cpp [hunk 3]
```diff
@@ -1620,6 +1623,30 @@ bool Compass::configured(char *failure_msg, uint8_t failure_msg_len)
...
+        }
+    }
+
+    // Lets do init one more time.
+    init();
```
Comment (tridge) [inline libraries/AP_Compass/AP_Compass.cpp:1646]: this calls the init functions again in the backends, which causes spurious errors and bus failures (eg. makes AK09916 on CubeOrange fail)

Comment (bugobliterator) [reply libraries/AP_Compass/AP_Compass.cpp:1646]: @tridge register_periodic_callback should not be called again, as calling register_compass method inside init calls second time should fail. I am trying and see if I can reproduce this.

File: libraries/AP_Compass/AP_Compass.cpp (hunk 2 above)
Comment (bugobliterator) [reply libraries/AP_Compass/AP_Compass.cpp:1641]: Yeah, that's a bug. Fixed

File: libraries/AP_Compass/AP_Compass.cpp (hunk 3 above)
Comment (bugobliterator) [reply libraries/AP_Compass/AP_Compass.cpp]: This issue was actually due to failure to reinitialise auxiliary bus over SPI. I have modified the driver to only try reinitialise external I2C and UAVCAN compasses. I have tested using Here Compass in I2C and UAVCAN mode. When we are going through I2C reinit the probe is not called for device with id already used, and similar mechanism is in place for UAVCAN as well. Also on top of that I have added code in register compass to return false if the call is being done for dev_id already registered.

**Comment by bugobliterator** [comment]
Scenarios Tested:

1. 2 Internal SPI compasses, and 1 external UAVCAN compass (CubeBlack):
    - Remove external UAVCAN compass, without any replacement connected
       Behaviour: arming check fails with compass missing. To make this go away calibration is required or manually setting the respective PRIO to 0
    - Both of the compasses were connected before a boot and then the one used is removed before next boot.
       Behaviour: 2nd external compass is recorded as unregistered additional compass, and arming check simply fails with compass missing. If user continues to calibrate, only two internal compasses will be calibrated. And unless user specifically moves the additional 2nd compass to the PRIO list it will never be used.
    - Remove the additional external UAVCAN compass, connect the new compass and reboot.
       Behaviour: 2nd external compass is now correctly detected as a replacement compass, and replaces the older compass. The user can continue to successfully calibrate all three compasses.

2. 1 Internal SPI compass and 2 external UAVCAN compasses (CubeOrange):
    - Both compasses connected at the same time and calibrated
    Behaviour: Both compasses will be allocated PRIO as there is enough empty space. Removing a compass will not lead to replacement by the other compass, and will be treated as a missing compass. The removed compass's PRIO will be reset when calibration is initiated, but not modification will take place other than removing gaps on next reboot.
   - Only one compass connected, and calibrated. That compass is removed, and second one is connected before a boot.
    Behaviour: The second compass will be correctly detected as a replacement compass and will replace the PRIO of first compass.

@tridge @rmackay9 I believe I have covered variations we discussed, let me know if you think I missed to cover a case.

**Comment by rmackay9** [comment]
This is included in Copter-4.0.4-rc3

**Pull request merged**